clang/lib/CConv/DeclRewriter.cpp

```diff
@@ -0,0 +1,410 @@
+          E->printPretty(NewMlDecl, nullptr, A.getPrintingPolicy());
+        } else {
+          if (isPointerType(VDL))
+            NewMlDecl << " = ((void *)0)";
+        }
+        NewMlDecl << ";\n";
+      } else {
+        DL->print(NewMlDecl);
+        NewMlDecl << ";\n";
+      }
+    }
+
+    // Step 4: Write out the string built up in step 3.
+    R.InsertTextAfter(EndOfLine, NewMlDecl.str());
+
+    // Step 5: Be sure and skip all of the NewTyps that we dealt with
+    //         during this time of hacking, by adding them to the
+    //         skip set.
+
+    for (const auto &TN : RewritesForThisDecl)
+      Skip.insert(TN);
+  } else {
+    if (Verbose) {
+      errs() << "Don't know how to re-write VarDecl\n";
+      VD->dump();
+      errs() << "at\n";
+      if (N.Statement)
+        N.Statement->dump();
+      errs() << "with " << N.Replacement << "\n";
+    }
+  }
+}
+
+void DeclRewriter::rewriteFunctionDecl(const DAndReplace &N) {
+  // TODO: If the return type is a fully-specified function pointer,
+  //       then clang will give back an invalid source range for the
+  //       return type source range. For now, check that the source
+  //       range is valid.
+  //       Additionally, a source range can be (mis) identified as
+  //       spanning multiple files. We don't know how to re-write that,
+  //       so don't.
+
+  FunctionDecl *UD = N.getDecl<FunctionDecl>();
+  SourceRange SR;
+  if (N.FullDecl) {
+    SR = UD->getSourceRange();
+    SR.setEnd(getFunctionDeclarationEnd(UD, A.getSourceManager()));
+  } else {
+    SR = UD->getReturnTypeSourceRange();
+  }
+  if (canRewrite(R, SR))
+    R.ReplaceText(SR, N.Replacement);
+}
+
+void DeclRewriter::rewrite(RSet &ToRewrite, std::set<FileID> &TouchedFiles) {
+  for (const auto &N : ToRewrite) {
+    assert(N.Declaration != nullptr);
+
+    if (Verbose) {
+      errs() << "Replacing type of decl:\n";
+      N.Declaration->dump();
+      errs() << "with " << N.Replacement << "\n";
+    }
+
+    // Get a FullSourceLoc for the start location and add it to the
+    // list of file ID's we've touched.
+    SourceRange tTR = N.Declaration->getSourceRange();
+    FullSourceLoc tFSL(tTR.getBegin(), A.getSourceManager());
+    TouchedFiles.insert(tFSL.getFileID());
+
+    // Exact rewriting procedure depends on declaration type
+    if (N.hasDeclType<ParmVarDecl>()) {
+      // TODO: why is this asserted?
+      assert(N.Statement == nullptr);
+      rewriteParmVarDecl(N);
+    } else if (N.hasDeclType<VarDecl>()) {
+      rewriteVarDecl(N, ToRewrite);
+    } else if (N.hasDeclType<FunctionDecl>()) {
+      rewriteFunctionDecl(N);
+    } else if (N.hasDeclType<FieldDecl>()) {
+      SourceRange SR = N.getDecl<FieldDecl>()->getSourceRange();
+      if (canRewrite(R, SR))
+        R.ReplaceText(SR, N.Replacement);
+    }
+  }
+}
+
+std::map<std::string, std::string> DeclRewriter::NewFuncSig;
+
+void DeclRewriter::rewriteDecls(ASTContext &Context, ProgramInfo &Info,
```

Comment:
Add a comment to remind the reader that this is the entry-point method for this class? And: Is it possible to put this method first, and then to read the code top-down? I would imagine that having declared the methods in the header file this is OK. Not sure what the rules are, stylistically.